### ISSUE TITLE
recursive calls in inline functions generates lots of code

### ISSUE THREAD

**Issue by romamik** [HaxeFoundation/haxe#7569]
After updating to haxe 4 preview 5, we noticed that our generated js became 3 time bigger: 10 mb to 30 mb. We found that this was because some of our inline functions were calling themselves recursively. With haxe 3.4 recursive call was not inlined, but haxe 4 preview 5 inlines some levels of recursion before making actual recursive call. 

For example

    class Test {
        static inline function test(counter: Int) {
          trace(counter);
          if(counter > 0) {
            test(counter - 1);
          }
        }
        static function main() {
            test(100);
            trace('ok');
        }
    }

with haxe 3.4 compiles to:
    
    // Generated by Haxe 3.4.4
    (function () { "use strict";
    var Test = function() { };
    Test.test = function(counter) {
      console.log(counter);
      if(counter > 0) {
        Test.test(counter - 1);
      }
    };
    Test.main = function() {
      console.log(100);
      Test.test(99);
      console.log("ok");
    };
    Test.main();
    })();
    
And with haxe 4 preview 5 to:
    
    // Generated by Haxe 4.0.0-preview.5+b260f99
    (function () { "use strict";
    var Test = function() { };
    Test.test = function(counter) {
      console.log("root/program/Test.hx:4:",counter);
      if(counter > 0) {
        var counter1 = counter - 1;
        console.log("root/program/Test.hx:4:",counter1);
        if(counter1 > 0) {
          var counter2 = counter1 - 1;
          console.log("root/program/Test.hx:4:",counter2);
          if(counter2 > 0) {
            Test.test(counter2 - 1);
          }
        }
      }
    };
    Test.main = function() {
      console.log("root/program/Test.hx:4:",100);
      console.log("root/program/Test.hx:4:",99);
      console.log("root/program/Test.hx:4:",98);
      console.log("root/program/Test.hx:4:",97);
      console.log("root/program/Test.hx:4:",96);
      console.log("root/program/Test.hx:4:",95);
      console.log("root/program/Test.hx:4:",94);
      Test.test(93);
      console.log("root/program/Test.hx:12:","ok");
    };
    Test.main();
    })();

**Comment by Simn** [comment]
I don't really understand how to avoid this, and why it didn't happen in the past. The initial inlining during typing is skipped because at that point the field has no expression yet (because we're currently typing it). However, during the optimization stage, it's just a normal call which the optimizer figures can be inlined.

We can detect this while optimizing the `test` field and avoid the inlining there, but then we still get multiple inlines from the `main` expression.

**Comment by Simn** [comment]
We now get this:

```js
Main.test = function(counter) {
	haxe_Log.trace(counter,{ fileName : "source/Main.hx", lineNumber : 3, className : "Main", methodName : "test"});
	if(counter > 0) {
		Main.test(counter - 1);
	}
};
Main.main = function() {
	haxe_Log.trace(100,{ fileName : "source/Main.hx", lineNumber : 3, className : "Main", methodName : "test"});
	haxe_Log.trace(99,{ fileName : "source/Main.hx", lineNumber : 3, className : "Main", methodName : "test"});
	Main.test(98);
	haxe_Log.trace("ok",{ fileName : "source/Main.hx", lineNumber : 10, className : "Main", methodName : "main"});
};
```

There are still two instances of `test` being inlined:

1. The 100 is the initial inline call that should be inlined.
2. The 99 is picked up during the post-process optimization because all the optimizer sees there is a call to `inline function test` being made from within the `main` expression.

Preventing 2. would require to tag the expression somehow. I guess we could inject some metadata and hope that it doesn't break any pattern matching.

**Comment by Simn** [comment]
I still don't really understand why this behavior changed given that Haxe 3 also had this post-processing optimization. The only thing I can think of is that the inlining got canceled for some random reason and that no longer happens with the recent improvements.

I mention this because I still struggle to come up with a solution here. We would want to not inline an inline call of a field f if it originates from inlining that same field f before. This would have to use a proper stack check because this kind of inline behavior can also occur with mutually recursive functions.

Basically, when inlining a call to `SomeClass.someField` we would have to leave a reference to that field in the AST. Then the post processing could pick it up and put it on the recursion check to avoid inlining.

**Comment by ncannasse** [comment]
I'm not sure too what to do here. It seems acceptable to me that inline recursive functions get inlined twice (as long as it's not infinite). @Simn feel free to move to Backlog

Is there a valid case were you want the function to be inlined but not the recursive call itself ?

**Comment by romamik** [comment]
IMHO recursive calls of inline functions is a bad practice. Bad thing is this can be done unintentionally, by mistake, especially in case of mutually recursive functions. And programmer can be unaware of what he done. In our case that had lead to very huge resulting code growth that was not mentioned by anybody for some time. Also it was not trivial to find places where recursive inlining has happened. 
For me current behavior is ok, but warning about recursive inlining would be great.

**Comment by Simn** [comment]
Thanks to TRE we now get this:

```js
Main.test = function(counter) {
	while(true) {
		console.log("source/Main.hx:3:",counter);
		if(counter > 0) {
			counter -= 1;
			continue;
		}
		return;
	}
};
Main.main = function() {
	console.log("source/Main.hx:3:",100);
	Main.test(99);
	console.log("source/Main.hx:11:","ok");
};
```

It's pretty silly to have the `100` as separate code here, but that's again not very surprising. The issue will go away if we address #8942.
